### ISSUE TITLE
Ensure GwtScriptOnly works

### ISSUE THREAD

**Issue by stephenh** [scalagwt/scalagwt-gwt#19]
Allows emulated code to be conditionally used in dev mode.

**Comment by gkossakowski** [comment]
Any examples how to use this and why it's useful? I tried to google GwtScriptOnly but found nothing enlightening.

Also, why this can't work with current implementation?

**Comment by stephenh** [comment]
If you search for GwtScriptOnly in the GWT repository, it's used in a few places: SafeStylesHostedModeUtils, SafeHtmlHostedModeUtils, ClientSerializationStreamReader.

All of these are super-sourced implementations of functionality that GWT (typically) uses for both dev mode and compile mode.

However, if dev mode's CompilingClassLoader sees GwtScriptOnly on a class, it ignores the super-sourced implementation and falls back to the regular JVM version. AFAIK this is typically for performance reasons.

E.g. GWT-RPC serialization/deserialization in dev mode used to be slow because it had to assume the user might be in compile mode and so it relied on Javascript parsing that was a lot of back and forth between the JVM and JS. However, if GWT-RPC can cheat and know that it really is running in the JVM, it can use a JVM-only implementation with litte/no trips to JS and so is a lot faster.

Beyond that example, I would have to defer to the GWT team on whether it is merely for perf optimization or for more than just that.

So, whether this works depends on how you're invoking DevMode--if you invoke it with a "regular" Java classpath (no gwt-lang.jar, or other hand-made GWT-only jars), then CompilingClassLoader should be able to choose the right class file as needed (the one that came from the GWT-specific compilation or the regular Java version).

Since the current GWT patch assumes the .class files on the classpath are the GWT-specific versions, there is no place to fallback to to get the pure-JVM versions of classes.

**Comment by gkossakowski** [comment]
> So, whether this works depends on how you're invoking DevMode--if you invoke it with a "regular" Java classpath (no gwt-lang.jar, or other hand-made GWT-only jars), then CompilingClassLoader should be able to choose the right class file as needed (the one that came from the GWT-specific compilation or the regular Java version).
> 
> Since the current GWT patch assumes the .class files on the classpath are the GWT-specific versions, there is no place to fallback to to get the pure-JVM versions of classes.

Not really. With my latest fixes to samples you can see that in [build.xml](https://github.com/scalagwt/scalagwt-sample/blob/master/build.xml) that `project.class.path` is defined in following way:

``` xml
  <path id="project.class.path">
    <pathelement location="war/WEB-INF/classes"/>
    <fileset dir="lib/gwt" includes="*.jar"/>
    <pathelement location="lib/scala/scala-library-gwt.jar"/>
    <!-- Add any additional non-server libs (such as JUnit) -->
    <fileset dir="war/WEB-INF/lib" includes="**/*.jar"/>
  </path>
```

Only additional jar is `scala-library-gwt.jar` which **doesn't include any gwt class files**. Class files in this jar are only for Scala library. It's true that in order to build `scala-library-gwt.jar` itself one needs `gwt-lang.jar` but this is required just to make sure that class files generated for library do not depend on classes/method gwt doesn't support.

In short: we run devmode with standard classpath and `scala-library-gwt.jar` is seen as any other application (client) code.

**Comment by stephenh** [comment]
Yes, if this pattern is followed, then this will work for any non-scala classes.

**Comment by gkossakowski** [comment]
Now the question naturally arises: why it won't work for Scala classes? :-)

**Comment by stephenh** [comment]
Because the "load .class files for the TypeOracle from the classpath" means there is only one place to get a class file from (and it's only the client-side version).

With ecj, GWT can choose between the (client-side) class file generated by ecj and the (server-side) class file that's on the classpath. E.g. for GwtScriptOnly to work for, say, a class in the scala-library, scala-library.jar would need to be on the classpath, and the scala-library-gwt.jar class files loaded ...some other way (for ecj, this is either directly from ecj after it compiles source files, or from gwtar files, or from the unit cache).

**Comment by gkossakowski** [comment]
Ah, you are right! Thanks for your explanation.

Since this feature seems to be used only in gwt internals (I failed to find any mention of it in official documentation) and we are not breaking it for any Java code I think it's reasonable to close this ticket. WDYT?

**Comment by stephenh** [comment]
Yeah, that's fine.
